data/transactions/logic/evalAppTxn_test.go

```diff
@@ -1140,3 +1170,30 @@ int 890
 ==
 `, ep)
 }
+
+// TestCallerGlobals checks that a called app can see its caller.
+func TestCallerGlobals(t *testing.T) {
+	ep, tx, ledger := makeSampleEnv()
+	globals := testProg(t, fmt.Sprintf(`
+global CallerApplicationID
+int 888
+==
+global CallerApplicationAddress
+addr %s
+==
+&&
+`, basics.AppIndex(888).Address()), AssemblerMaxVersion)
+	ledger.NewApp(tx.Receiver, 222, basics.AppParams{
+		ApprovalProgram: globals.Program,
+	})
+
+	ledger.NewApp(tx.Receiver, 888, basics.AppParams{})
+	ledger.NewAccount(ledger.ApplicationID().Address(), 50_000)
+	tx.ForeignApps = []basics.AppIndex{basics.AppIndex(222)}
+	testApp(t, `itxn_begin
+int appl;    itxn_field TypeEnum
+int 222;     itxn_field ApplicationID
```

Comment:
Interesting -- so in this test, `NewApp` only creates the app with id 222 (but not run), and the actual TEAL code is run right after the `itxn_submit` in the `testApp` function? 

I also probably have to read deeper into the original C2C PR, but I'm also wondering how inner contracts fail/error (do they just propagate the errors back up using the `EvalParams`?)

Reply:
Correct, for unit tests in the logic package, "NewApp" sets up an "app" in a fake ledger (ledger_test.go).  Then it is invoked as an inner app call by the code in the "main" app. In more realistic tests it's slightly trickier because the only way to install that first app is by making an app call transaction and finding out the app id that it gets assigned. (Also, writing that app must ensure it does nothing during the create call.)

If an inner app fails, the failure cascades and everything is torn down and fails, no matter how deep. This is what happens in the case of any inner txn failure, so there's no special case for apps. See opTxSubmit in logic/eval.go